Userland/Applications/PixelPaint/Tools/GradientTool.cpp

```diff
@@ -346,32 +441,65 @@ void GradientTool::reset()
     m_gradient_start = {};
     m_gradient_center = {};
     m_gradient_end = {};
+    m_gradient_transversal_a = {};
+    m_gradient_transversal_b = {};
     m_gradient_half_length = 0;
     m_physical_diagonal_layer_length = 0;
     m_hover_over_drag_handle = false;
     m_hover_over_start_handle = false;
     m_hover_over_end_handle = false;
+    m_hover_over_transversal_a_handle = false;
+    m_hover_over_transversal_b_handle = false;
 
     if (m_editor) {
         m_editor->update();
         m_editor->update_tool_cursor();
     }
 }
 
-void GradientTool::update_gradient_end_and_derive_start(Gfx::IntPoint const new_end_point)
+void GradientTool::update_gradient_with_initial_values(Gfx::IntPoint const new_end_point)
 {
     VERIFY(m_gradient_center.has_value());
     m_gradient_end = new_end_point;
-    m_gradient_start = m_gradient_center.value() - (m_gradient_end.value() - m_gradient_center.value());
+    auto deltaCenter = m_gradient_end.value() - m_gradient_center.value();
+    m_gradient_start = m_gradient_center.value() - deltaCenter;
+
+    if (m_mode == GradientMode::Radial) {
+        Gfx::IntPoint perpendicularDeltaCenter = { -deltaCenter.y(), deltaCenter.x() };
+        m_gradient_transversal_a = m_gradient_center.value() + perpendicularDeltaCenter;
+        m_gradient_transversal_b = m_gradient_center.value() - perpendicularDeltaCenter;
+    }
+}
+
+void GradientTool::move_gradient_position(Gfx::IntPoint const movement_delta)
+{
+    m_gradient_end.value().translate_by(movement_delta);
+    m_gradient_start.value().translate_by(movement_delta);
+
+    if (m_mode == GradientMode::Radial) {
+        m_gradient_transversal_a.value().translate_by(movement_delta);
+        m_gradient_transversal_b.value().translate_by(movement_delta);
+    }
 }
 
-void GradientTool::translate_gradient_start_end(Gfx::IntPoint const delta, bool update_start_counterwise)
+void GradientTool::rotate_gradient_points(Gfx::IntPoint const delta)
 {
     m_gradient_end.value().translate_by(delta);
-    if (update_start_counterwise)
-        m_gradient_start.value().translate_by(delta.scaled(-1, -1));
-    else
-        m_gradient_start.value().translate_by(delta);
+    auto translation_distance_to_center = m_gradient_center.value().distance_from(m_gradient_end.value()) - m_gradient_center.value().distance_from(m_gradient_start.value());
+    m_gradient_start.value().translate_by(delta.scaled(-1, -1));
+
+    if (m_mode == GradientMode::Radial) {
+        auto new_horizontal_distance_fraction = (translation_distance_to_center + m_gradient_center.value().distance_from(m_gradient_transversal_a.value())) / m_gradient_center.value().distance_from(m_gradient_start.value());
+        calculate_transversal_points(new_horizontal_distance_fraction);
+    }
+}
+
+void GradientTool::calculate_transversal_points(float scale_fraction)
```

Comment:
based on the parameter fed above iʼd guess that `scale_fraction` is in fact an “eccentricity”… care to confirm that?